File: internal/seed/football.go
package seed

import (
	"context"
	"fmt"
	"log/slog"

	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/albapepper/scoracle-data/internal/provider"
	"github.com/albapepper/scoracle-data/internal/provider/sportmonks"
)

const sportFootball = "FOOTBALL"

// SeedFootballSeason seeds all data for a single Football league-season.
//
// Args:
//   - smSeasonID: SportMonks season ID (from provider_seasons table)
//   - leagueID: Our internal league ID (8, 82, 301, 384, 564)
//   - seasonYear: Year (e.g. 2024 for 2024-25 season)
//   - smLeagueID: SportMonks league ID (same as our internal ID for football)
func SeedFootballSeason(
	ctx context.Context,
	pool *pgxpool.Pool,
	handler *sportmonks.FootballHandler,
	smSeasonID int,
	leagueID int,
	seasonYear int,
	smLeagueID int,
	logger *slog.Logger,
) SeedResult {
	var result SeedResult

	// Resolve SportMonks league ID from DB if not provided
	if smLeagueID == 0 {
		var dbSmID *int
		var leagueName string
		err := pool.QueryRow(ctx, "league_lookup", leagueID).Scan(&dbSmID, &leagueName)
		if err != nil || dbSmID == nil {
			result.AddErrorf("no sportmonks_id found for league %d: %v", leagueID, err)
			return result
		}
		smLeagueID = *dbSmID
		logger.Info("Resolved SportMonks league", "league_id", leagueID, "name", leagueName, "sm_id", smLeagueID)
	}

	logger.Info("Seeding football season",
		"sm_season_id", smSeasonID, "league_id", leagueID,
		"season_year", seasonYear, "sm_league_id", smLeagueID)

	// 1. Teams
	logger.Info("Phase 1/3: Seeding teams...")
	teams, err := handler.GetTeams(ctx, smSeasonID)
	if err != nil {
		result.AddErrorf("fetch teams: %v", err)
	} else {
		for _, team := range teams {
			if err := UpsertTeam(ctx, pool, sportFootball, team); err != nil {
				result.AddErrorf("upsert team %d: %v", team.ID, err)
			} else {
				result.TeamsUpserted++
			}
		}
	}
	logger.Info("Teams done", "count", result.TeamsUpserted)

	// 2. Players + Player Stats (via squad iteration)
	logger.Info("Phase 2/3: Seeding players + stats...")
	teamIDs := make([]int, len(teams))
	for i, t := range teams {
		teamIDs[i] = t.ID
	}

	count := 0
	err = handler.GetPlayersWithStats(ctx, smSeasonID, teamIDs, smLeagueID,
		func(ps provider.PlayerStats) error {
			if ps.Player != nil {
				if err := UpsertPlayer(ctx, pool, sportFootball, *ps.Player); err != nil {
					result.AddErrorf("upsert player %d: %v", ps.PlayerID, err)
				} else {
					result.PlayersUpserted++
				}
			}
			if len(ps.Stats) > 0 {
				if err := UpsertPlayerStats(ctx, pool, sportFootball, seasonYear, leagueID, ps); err != nil {
					result.AddErrorf("upsert player stats %d: %v", ps.PlayerID, err)
				} else {
					result.PlayerStatsUpserted++
				}
			}
			count++
			if count%50 == 0 {
				logger.Info("Player progress", "count", count)
			}
			return nil
		})
	if err != nil {
		result.AddErrorf("fetch players/stats: %v", err)
	}
	logger.Info("Players + stats done",
		"players", result.PlayersUpserted, "stats", result.PlayerStatsUpserted)

	// 3. Team Stats (Standings)
	logger.Info("Phase 3/3: Seeding standings...")
	teamStats, err := handler.GetTeamStats(ctx, smSeasonID)
	if err != nil {
		result.AddErrorf("fetch standings: %v", err)
	} else {
		for _, ts := range teamStats {
			if ts.Team != nil {
				if err := UpsertTeam(ctx, pool, sportFootball, *ts.Team); err != nil {
					result.AddErrorf("upsert team from standings %d: %v", ts.TeamID, err)
				}
			}
			if err := UpsertTeamStats(ctx, pool, sportFootball, seasonYear, leagueID, ts); err != nil {
				result.AddErrorf("upsert team stats %d: %v", ts.TeamID, err)
			} else {
				result.TeamStatsUpserted++
			}
		}
	}
	logger.Info("Standings done", "count", result.TeamStatsUpserted)

	logger.Info("Football season seed complete",
		"league_id", leagueID, "season", seasonYear, "summary", result.Summary())
	return result
}

// ResolveProviderSeasonID looks up the SportMonks season ID from the provider_seasons table.
func ResolveProviderSeasonID(ctx context.Context, pool *pgxpool.Pool, leagueID, seasonYear int) (int, error) {
	var smSeasonID *int
	err := pool.QueryRow(ctx, "resolve_provider_season", leagueID, seasonYear).Scan(&smSeasonID)
	if err != nil {
		return 0, fmt.Errorf("resolve provider season: %w", err)
	}
	if smSeasonID == nil {
		return 0, fmt.Errorf("no provider season found for league %d season %d", leagueID, seasonYear)
	}
	return *smSeasonID, nil
}

File: legacy_fastapi/python/scoracle_data/seeders/football.py
"""
Football seed runner — league/team iteration for SportMonks data.

NBA/NFL use BaseSeedRunner directly (identical flow: teams -> players -> stats).
Football differs because:
  - Data is organized per-league (Premier League, La Liga, etc.)
  - Player stats require squad-by-squad iteration (bulk endpoint empty on our tier)
  - Season IDs must be resolved from provider_seasons table

All API calls and normalization are handled by SportMonksHandler.
This file only orchestrates the iteration order and DB writes.
"""

from __future__ import annotations

import logging
from typing import Any, TYPE_CHECKING

from .base import BaseSeedRunner
from .common import SeedResult

if TYPE_CHECKING:
    from ..handlers.sportmonks import SportMonksHandler
    from ..pg_connection import PostgresDB

logger = logging.getLogger(__name__)

SPORT = "FOOTBALL"


class FootballSeedRunner(BaseSeedRunner):
    """Seeds Football data with per-league, per-team iteration.

    Unlike NBA/NFL (which use BaseSeedRunner.seed_all directly),
    Football requires seed_season() which resolves SportMonks IDs,
    iterates teams for squad-based player stats, and handles standings.
    """

    def __init__(self, db: "PostgresDB", handler: "SportMonksHandler"):
        super().__init__(db, handler, SPORT)

    async def seed_season(
        self,
        sm_season_id: int,
        league_id: int,
        season_year: int,
        sportmonks_league_id: int | None = None,
    ) -> SeedResult:
        """Seed all data for a single Football league-season.

        Args:
            sm_season_id: SportMonks season ID (from provider_seasons table).
            league_id: Our internal league ID (1-5).
            season_year: Year (e.g. 2024 for 2024-25 season).
            sportmonks_league_id: SportMonks league ID (auto-resolved if not provided).
        """
        # Resolve sportmonks_league_id if not provided
        if sportmonks_league_id is None:
            league_row = self.db.fetchone(
                "SELECT sportmonks_id, name FROM leagues WHERE id = %s",
                (league_id,),
            )
            if not league_row or not league_row["sportmonks_id"]:
                raise ValueError(f"No sportmonks_id found for league {league_id}")
            sportmonks_league_id = int(league_row["sportmonks_id"])
            league_name = league_row["name"]
        else:
            league_row = self.db.fetchone(
                "SELECT name FROM leagues WHERE id = %s",
                (league_id,),
            )
            league_name = league_row["name"] if league_row else f"League {league_id}"

        logger.info(
            "Seeding season %d (league %d: %s, year %d)",
            sm_season_id, league_id, league_name, season_year,
        )

        result = SeedResult()
        teams: list[dict] = []

        # 1. Teams
        logger.info("Phase 1/3: Seeding teams...")
        try:
            teams = await self.handler.get_teams(sm_season_id)
            for team in teams:
                self._upsert_team(team)
                result.teams_upserted += 1
            logger.info("Upserted %d teams", result.teams_upserted)
        except Exception as e:
            error_msg = f"Error seeding teams: {e}"
            logger.error(error_msg)
            result.errors.append(error_msg)

        # 2. Players + Player Stats (fetched together via squad iteration)
        logger.info("Phase 2/3: Seeding players + player stats...")
        try:
            team_ids = [t["id"] for t in teams] if result.teams_upserted > 0 else []
            count = 0
            async for data in self.handler.get_players_with_stats(
                sm_season_id, team_ids, sportmonks_league_id
            ):
                try:
                    # Upsert player profile
                    if data.get("player"):
                        self._upsert_player(data["player"])
                        result.players_upserted += 1
                    # Upsert player stats
                    if data.get("stats"):
                        self._upsert_player_stats(data, season_year, league_id)
                        result.player_stats_upserted += 1
                    count += 1
                    if count % 50 == 0:
                        logger.info("Processed %d players...", count)
                except Exception as e:
                    pid = data.get("player_id", "?")
                    result.errors.append(f"Error upserting player {pid}: {e}")
        except Exception as e:
            error_msg = f"Error seeding players/stats: {e}"
            logger.error(error_msg)
            result.errors.append(error_msg)

        # 3. Team Stats (Standings)
        logger.info("Phase 3/3: Seeding team stats (standings)...")
        try:
            team_stats = await self.handler.get_team_stats(sm_season_id)
            for data in team_stats:
                try:
                    # Auto-upsert team if handler provides it
                    if data.get("team"):
                        self._upsert_team(data["team"])
                    self._upsert_team_stats(data, season_year, league_id)
                    result.team_stats_upserted += 1
                except Exception as e:
                    tid = data.get("team_id", "?")
                    result.errors.append(f"Error upserting team stats for {tid}: {e}")
            logger.info("Upserted %d team stats", result.team_stats_upserted)
        except Exception as e:
            error_msg = f"Error seeding team stats: {e}"
            logger.error(error_msg)
            result.errors.append(error_msg)

        logger.info(
            "Season seeding complete: %d teams, %d players, "
            "%d player stats, %d team stats, %d errors",
            result.teams_upserted,
            result.players_upserted,
            result.player_stats_upserted,
            result.team_stats_upserted,
            len(result.errors),
        )

        return result
